File: blf_lib/src/blam/haloreach/v09730_10_04_09_1309_omaha_delta/game/game_engine_team.rs
use num_derive::{FromPrimitive, ToPrimitive};
use num_traits::FromPrimitive;
use serde::{Deserialize, Serialize};
use blf_lib::io::bitstream::{c_bitstream_reader, c_bitstream_writer};
use blf_lib::blam::haloreach::v12065_11_08_24_1738_tu1actual::game::string_table::c_string_table;
use blf_lib_derivable::result::BLFLibResult;
use crate::types::array::StaticArray;
use serde_hex::{SerHex,StrictCap};

pub const k_game_variant_team_count: usize = 8;

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default, ToPrimitive, FromPrimitive, crate::derive::c_enum)]
#[bits(3)]
pub enum e_game_engine_team_options_model_override_type {
    #[default]
    none = 0,
    spartan = 1,
    elite = 2,
    set_by_team = 3,
    by_designator = 4,
}

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default, ToPrimitive, FromPrimitive, crate::derive::c_enum)]
#[bits(2)]
pub enum e_game_engine_team_options_designator_switch_type {
    #[default]
    none = 0,
    random = 1,
    rotate = 2,
}

#[repr(i8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default, ToPrimitive, FromPrimitive, crate::derive::c_enum)]
#[bits(4)]
pub enum e_multiplayer_team_designator {
    #[default]
    none = -1,
    defenders = 0,
    attackers = 1,
    third_party = 2,
    fourth_party = 3,
    fifth_party = 4,
    sixth_party = 5,
    seventh_party = 6,
    eighth_party = 7,
    neutral = 8,
}

#[repr(i8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default, ToPrimitive, FromPrimitive, crate::derive::c_enum)]
#[bits(1)]
pub enum e_player_model_choice {
    #[default]
    spartan = 0,
    elite = 1,
}

#[derive(Default, PartialEq, Debug, Clone, Serialize, Deserialize)]
pub struct c_game_engine_team_options_team {
    pub m_team_enabled: bool,
    pub m_override_color_armour: bool,
    pub m_override_color_ui_text: bool,
    pub m_override_color_ui_bitmap: bool,
    pub m_name: c_string_table<1, 32, 5, 6, 1>,
    pub m_team_initial_designator: e_multiplayer_team_designator,
    pub m_model_override: e_player_model_choice,
    #[serde(with = "SerHex::<StrictCap>")]
    pub m_team_color_override: u32,
    #[serde(with = "SerHex::<StrictCap>")]
    pub m_team_ui_text_tint_color_override: u32,
    #[serde(with = "SerHex::<StrictCap>")]
    pub m_team_ui_bitmap_tint_color_override: u32,
}

impl c_game_engine_team_options_team {
    pub fn initialize(&mut self, team_index: usize) {
        *self = Self::default();
        self.m_team_enabled = true;
        self.m_team_color_override = 0xFFFF_FFFF;
        self.m_team_ui_text_tint_color_override = 0xFFFF_FFFF;
        self.m_team_ui_bitmap_tint_color_override = 0xFFFF_FFFF;
        self.m_team_initial_designator = e_multiplayer_team_designator::from_i8(team_index as i8 + 1)
            .unwrap_or(e_multiplayer_team_designator::none);
    }

    pub fn encode(&self, bitstream: &mut c_bitstream_writer) -> BLFLibResult {
        bitstream.write_bool(self.m_override_color_ui_bitmap)?;
        bitstream.write_bool(self.m_override_color_ui_text)?;
        bitstream.write_bool(self.m_override_color_armour)?;
        bitstream.write_bool(self.m_team_enabled)?;
        self.m_name.encode(bitstream)?;
        bitstream.write_enum(self.m_team_initial_designator)?;
        bitstream.write_enum(self.m_model_override)?;
        bitstream.write_integer(self.m_team_color_override, 32)?;
        bitstream.write_integer(self.m_team_ui_text_tint_color_override, 32)?;
        bitstream.write_integer(self.m_team_ui_bitmap_tint_color_override, 32)?;

        Ok(())
    }

    pub fn decode(&mut self, bitstream: &mut c_bitstream_reader) -> BLFLibResult {
        self.m_override_color_ui_bitmap = bitstream.read_bool("team-override-ui-bitmap-color")?;
        self.m_override_color_ui_text = bitstream.read_bool("team-override-ui-text-color")?;
        self.m_override_color_armour = bitstream.read_bool("team-override-armour-color")?;
        self.m_team_enabled = bitstream.read_bool("team-enabled")?;
        self.m_name.decode(bitstream)?;
        self.m_team_initial_designator = bitstream.read_enum("team-initial-designator")?;
        self.m_model_override = bitstream.read_enum("team-model-override")?;
        self.m_team_color_override = bitstream.read_integer("team-color-override", 32)?;
        self.m_team_ui_text_tint_color_override = bitstream.read_integer("team-ui-text-tint-color-override", 32)?;
        self.m_team_ui_bitmap_tint_color_override = bitstream.read_integer("team-ui-bitmap-tint-color-override", 32)?;

        Ok(())
    }
}


#[derive(Default, PartialEq, Debug, Clone, Serialize, Deserialize)]
pub struct c_game_engine_team_options {
    pub m_model_override: e_game_engine_team_options_model_override_type,
    pub m_designator_switch_type: e_game_engine_team_options_designator_switch_type,
    pub m_teams: StaticArray<c_game_engine_team_options_team, k_game_variant_team_count>,
}

impl c_game_engine_team_options {
    pub fn initialize(&mut self) {
        *self = Self::default();
        self.m_designator_switch_type =
            e_game_engine_team_options_designator_switch_type::rotate;
        for i in 0..k_game_variant_team_count {
            self.m_teams[i].initialize(i);
        }
    }

    pub fn encode(&self, bitstream: &mut c_bitstream_writer) -> BLFLibResult {
        bitstream.write_enum(self.m_model_override)?;
        bitstream.write_enum(self.m_designator_switch_type)?;
        for i in 0..k_game_variant_team_count {
            self.m_teams[i].encode(bitstream)?
        }

        Ok(())
    }

    pub fn decode(&mut self, bitstream: &mut c_bitstream_reader) -> BLFLibResult {
        self.m_model_override = bitstream.read_enum("model-override")?;
        self.m_designator_switch_type = bitstream.read_enum("designator-switch-type")?;
        for team in self.m_teams.get_mut().iter_mut() {
            team.decode(bitstream)?
        }

        Ok(())
    }
}

File: blf-ts/src/blam/haloreach/v08516_10_02_19_1607_omaha_alpha/game/game_engine_team.ts
import type {
  c_bitstream_reader,
  c_bitstream_writer,
} from "../../../../bitstream";
import { AutoMap } from "../../../../helpers/automap";
import { type StaticArray, staticArray } from "../../../../types/static_array";
import {
  e_game_engine_team_options_designator_switch_type,
  e_multiplayer_team_designator,
  e_player_model_choice,
  k_game_variant_team_count,
} from "../../v12065_11_08_24_1738_tu1actual/game/game_engine_default";
import { c_string_table } from "../../v12065_11_08_24_1738_tu1actual/game/string_table";

/**
 * Matches blf_lib omaha_delta/alpha `c_game_engine_team_options_team`
 * (no fireteam_count — that arrives later in TU1).
 */
export class c_game_engine_team_options_team {
  @AutoMap(() => Boolean)
  m_team_enabled = false;
  @AutoMap(() => Boolean)
  m_override_color_armour = false;
  @AutoMap(() => Boolean)
  m_override_color_ui_text = false;
  @AutoMap(() => Boolean)
  m_override_color_ui_bitmap = false;
  @AutoMap(() => c_string_table)
  m_name = new c_string_table(1, 32, 5, 6, 1);
  @AutoMap(() => Number)
  m_team_initial_designator: e_multiplayer_team_designator =
    e_multiplayer_team_designator.none;
  @AutoMap(() => Number)
  m_model_override: e_player_model_choice = e_player_model_choice.spartan;
  @AutoMap(() => Number)
  m_team_color_override = 0;
  @AutoMap(() => Number)
  m_team_ui_text_tint_color_override = 0;
  @AutoMap(() => Number)
  m_team_ui_bitmap_tint_color_override = 0;
  initialize(team_index: number): void {
    this.m_team_enabled = true;
    this.m_override_color_armour = false;
    this.m_override_color_ui_text = false;
    this.m_override_color_ui_bitmap = false;
    this.m_name = new c_string_table(1, 32, 5, 6, 1);
    this.m_team_initial_designator = team_index;
    this.m_model_override = e_player_model_choice.spartan;
    this.m_team_color_override = 0xffffffff;
    this.m_team_ui_text_tint_color_override = 0xffffffff;
    this.m_team_ui_bitmap_tint_color_override = 0xffffffff;
  }
  decode(bitstream: c_bitstream_reader): void {
    this.m_override_color_ui_bitmap = bitstream.read_bool(
      "team-override-ui-bitmap-color"
    );
    this.m_override_color_ui_text = bitstream.read_bool(
      "team-override-ui-text-color"
    );
    this.m_override_color_armour = bitstream.read_bool(
      "team-override-armour-color"
    );
    this.m_team_enabled = bitstream.read_bool("team-enabled");
    this.m_name.decode(bitstream);
    this.m_team_initial_designator = bitstream.read_enum(
      "team-initial-designator",
      4,
      e_multiplayer_team_designator
    );
    this.m_model_override = bitstream.read_enum(
      "team-model-override",
      1,
      e_player_model_choice
    );
    this.m_team_color_override = bitstream.read_integer(
      "team-color-override",
      32
    );
    this.m_team_ui_text_tint_color_override = bitstream.read_integer(
      "team-ui-text-tint-color-override",
      32
    );
    this.m_team_ui_bitmap_tint_color_override = bitstream.read_integer(
      "team-ui-bitmap-tint-color-override",
      32
    );
  }
  encode(bitstream: c_bitstream_writer): void {
    bitstream.write_bool(this.m_override_color_ui_bitmap);
    bitstream.write_bool(this.m_override_color_ui_text);
    bitstream.write_bool(this.m_override_color_armour);
    bitstream.write_bool(this.m_team_enabled);
    this.m_name.encode(bitstream);
    bitstream.write_enum(
      this.m_team_initial_designator,
      4,
      e_multiplayer_team_designator
    );
    bitstream.write_enum(this.m_model_override, 1, e_player_model_choice);
    bitstream.write_integer(this.m_team_color_override, 32);
    bitstream.write_integer(this.m_team_ui_text_tint_color_override, 32);
    bitstream.write_integer(this.m_team_ui_bitmap_tint_color_override, 32);
  }
}

export class c_game_engine_team_options {
  @AutoMap(() => Number)
  m_model_override = 0;
  @AutoMap(() => e_game_engine_team_options_designator_switch_type)
  m_designator_switch_type: e_game_engine_team_options_designator_switch_type =
    e_game_engine_team_options_designator_switch_type.none;
  @AutoMap(() => [c_game_engine_team_options_team])
  m_teams: StaticArray<
    c_game_engine_team_options_team,
    typeof k_game_variant_team_count
  > = staticArray(
    k_game_variant_team_count,
    () => new c_game_engine_team_options_team()
  );
  initialize(): void {
    this.m_model_override = 0;
    this.m_designator_switch_type =
      e_game_engine_team_options_designator_switch_type.rotate;
    for (let i = 0; i < this.m_teams.length; i++) {
      this.m_teams[i]!.initialize(i);
    }
  }
  decode(bitstream: c_bitstream_reader): void {
    this.m_model_override = bitstream.read_integer("model-override", 3);
    this.m_designator_switch_type = bitstream.read_enum(
      "designator-switch-type",
      2,
      e_game_engine_team_options_designator_switch_type
    );
    for (const team of this.m_teams) {
      team.decode(bitstream);
    }
  }
  encode(bitstream: c_bitstream_writer): void {
    bitstream.write_integer(this.m_model_override, 3);
    bitstream.write_enum(
      this.m_designator_switch_type,
      2,
      e_game_engine_team_options_designator_switch_type
    );
    for (const team of this.m_teams) {
      team.encode(bitstream);
    }
  }
}
